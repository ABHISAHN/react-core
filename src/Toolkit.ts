export class Toolkit {
	static TESTING: boolean = false;
	static TESTING_UID = 0;

	/**
	 * Generates a unique ID (thanks Stack overflow :3)
	 */
	public static UID(): string {
		if (Toolkit.TESTING) {
			Toolkit.TESTING_UID++;
			return "" + Toolkit.TESTING_UID;
		}
		return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, c => {
			const r = (Math.random() * 16) | 0;
			const v = c === "x" ? r : (r & 0x3) | 0x8;
			return v.toString(16);
		});
	}
}
