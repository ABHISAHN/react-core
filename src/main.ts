export * from "./BaseObject";
export * from "./Toolkit";
export * from "./widgets/BaseWidget";
export * from "./widgets/MouseWidget";
