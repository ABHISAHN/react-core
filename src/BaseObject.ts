import { Toolkit } from "./Toolkit";

export interface BaseEvent<T extends BaseObject = any> {
	source: T;
	stopPropagation: () => any;
	firing: boolean;
	name: string;
}

export interface BaseListener<T extends BaseObject = any> {}

export class BaseObject<LISTENER extends BaseListener = BaseListener> {
	public listeners: { [s: string]: LISTENER };

	constructor() {
		this.listeners = {};
	}

	public iterateListeners(name: string, cb: (t: LISTENER, event: BaseEvent) => any) {
		let event: BaseEvent = {
			firing: true,
			source: this,
			name: name,
			stopPropagation: () => {
				event.firing = false;
			}
		};

		for (let i in this.listeners) {
			if (this.listeners.hasOwnProperty(i)) {
				// propagation stopped
				if (!event.firing) {
					return;
				}
				cb(this.listeners[i], event);
			}
		}
		return event;
	}

	public removeListener(listener: string) {
		if (this.listeners[listener]) {
			delete this.listeners[listener];
			return true;
		}
		return false;
	}

	public addListener(listener: LISTENER): string {
		let uid = Toolkit.UID();
		this.listeners[uid] = listener;
		return uid;
	}
}
