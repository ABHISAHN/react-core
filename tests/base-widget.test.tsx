import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { BaseWidget } from "../src/main";

class BaseWidgetTest extends BaseWidget{

	constructor(p){
		super("core-test",p);
	}

	render(){
		return (
			<div {...this.getProps()}>
				<div className={this.bem('__child1')} />
				<div className={this.bem(['__child2', '--modifier1'])} />
				<div className={this.bem({
					'--show-this-modifier': true,
					'--also-this-modifier': true,
					'--hide-this-modifier': false,
				})} />
			</div>
		);
	}
}

test('BaseWidget rendering', () => {
	const component = renderer.create(
		<BaseWidgetTest />,
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('BaseWidget rendering with extra props', () => {
	const component = renderer.create(
		<BaseWidgetTest extraProps={{style: {color: 'green'}}} />,
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('BaseWidget rendering with different base class', () => {
	const component = renderer.create(
		<BaseWidgetTest baseClass="core-test-override" />,
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('BaseWidget rendering with appended classNames', () => {
	const component = renderer.create(
		<BaseWidgetTest baseClass="core-test-override" className="extra-class" />,
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});