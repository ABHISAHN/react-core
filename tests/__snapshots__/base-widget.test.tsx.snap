// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseWidget rendering 1`] = `
<div
  className="core-test"
>
  <div
    className="core-test__child1"
  />
  <div
    className="core-test__child2 core-test--modifier1"
  />
  <div
    className="core-test--show-this-modifier core-test--also-this-modifier"
  />
</div>
`;

exports[`BaseWidget rendering with appended classNames 1`] = `
<div
  className="core-test-override extra-class"
>
  <div
    className="core-test-override__child1"
  />
  <div
    className="core-test-override__child2 core-test-override--modifier1"
  />
  <div
    className="core-test-override--show-this-modifier core-test-override--also-this-modifier"
  />
</div>
`;

exports[`BaseWidget rendering with different base class 1`] = `
<div
  className="core-test-override"
>
  <div
    className="core-test-override__child1"
  />
  <div
    className="core-test-override__child2 core-test-override--modifier1"
  />
  <div
    className="core-test-override--show-this-modifier core-test-override--also-this-modifier"
  />
</div>
`;

exports[`BaseWidget rendering with extra props 1`] = `
<div
  className="core-test"
  style={
    Object {
      "color": "green",
    }
  }
>
  <div
    className="core-test__child1"
  />
  <div
    className="core-test__child2 core-test--modifier1"
  />
  <div
    className="core-test--show-this-modifier core-test--also-this-modifier"
  />
</div>
`;
